fix(ngcc): handle entry-points within container folders

petebacondarwin opened the pull request:
The previous optimizations in #35756 to the
`DirectoryWalkerEntryPointFinder` were over zealous
with regard to packages that have entry-points stored
in "container" directories in the package, where the
container directory was not an entry-point itself.

Now we will also walk such "container" folders as long
as they do not contain `.js` files, which we regard as an
indicator that the directory will not contain entry-points.

Fixes #36216

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

petebacondarwin committed:
refactor(ngcc): simplify `DirectoryWalkerEntryPointFinder`

This commit simplifies the `DirectoryWalkerEntryPointFinder` inter-method
calling to make it easier to follow, and also to support controlling
walking of a directory based on its children.

alan-agius4 commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -160,9 +161,15 @@ export class DirectoryWalkerEntryPointFinder implements EntryPointFinder {
+            const childPaths = this.fs.readdir(absolutePath);
+            // Non-entry-point directories that contain JS files are very unlikely to be containers
+            // for sub-entry-points.
+            if (isEntryPoint || paths.every(path => !path.endsWith('.js'))) {
```
Wouldn’t this `!paths.some(path => path.endsWith('.js'))` be faster?

petebacondarwin replied:
Good spot!

gkalpak replied:
OOC, why would it be faster?

alan-agius4 replied:
I think in this case actually it doesn’t matter that much as most likely they will short-circuit in the same time.

alan-agius4 commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -140,31 +146,42 @@ export class DirectoryWalkerEntryPointFinder implements EntryPointFinder {
-          }
+      packagePath: AbsoluteFsPath, dir: AbsoluteFsPath, paths: PathSegment[],
+      checkForEntryPoint: (path: AbsoluteFsPath, isDirectory: boolean) => boolean): void {
+    return paths.forEach(path => {
```
- Redundant `return` 
- Also in such cases my personal preference is to use `for of` instead of `forEach`

petebacondarwin replied:
You caught me in a late night refactor :-)

alan-agius4 reviewed (approved):
One small NIT, other than that LGTM.

petebacondarwin commented:
I'm going to run a few performance measurements against master this morning just to see what impact this has.

alan-agius4 commented:
@petebacondarwin awesome, keep me posted :)

alan-agius4 commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -140,31 +146,42 @@ export class DirectoryWalkerEntryPointFinder implements EntryPointFinder {
+    for (const path of paths) {
+      if (path.startsWith('.') || path === 'node_modules' || path === NGCC_DIRECTORY) {
+        // Igore hidden files, node_modules and ngcc directory
+        return;
```
This now needs to be changed to `continue` to skip to the next iteration.

petebacondarwin replied:
Doh!

alan-agius4 commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -140,31 +146,42 @@ export class DirectoryWalkerEntryPointFinder implements EntryPointFinder {
+      const stat = this.fs.lstat(absolutePath);
+      if (stat.isSymbolicLink()) {
+        // We are not interested in symbolic links
+        return;
```
Same

alan-agius4 commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -140,31 +146,42 @@ export class DirectoryWalkerEntryPointFinder implements EntryPointFinder {
+
+      if (!isDirectory) {
+        // This path is not a directory so we are done
+        return;
```
Same

petebacondarwin committed:
refactor(ngcc): rename INVALID_ENTRY_POINT to INCOMPATIBLE_ENTRY_POINT

This name better reflects its meaning.

petebacondarwin committed:
fix(ngcc): handle entry-points within container folders

The previous optimizations in #35756 to the
`DirectoryWalkerEntryPointFinder` were over zealous
with regard to packages that have entry-points stored
in "container" directories in the package, where the
container directory was not an entry-point itself.

Now we will also walk such "container" folders as long
as they do not contain `.js` files, which we regard as an
indicator that the directory will not contain entry-points.

Fixes #36216

gkalpak commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -33,141 +33,150 @@ export class DirectoryWalkerEntryPointFinder implements EntryPointFinder {
-      if (subEntryPoint === NO_ENTRY_POINT || subEntryPoint === INVALID_ENTRY_POINT) {
-        return false;
+      if (subEntryPoint !== NO_ENTRY_POINT && subEntryPoint !== INCOMPATIBLE_ENTRY_POINT &&
+          subEntryPoint.compiledByAngular) {
```
Is this (`subEntryPoint.compiledByAngular`) a new check? Does it change the current behavior (in which case this is more than a refactoring :wink:)?

petebacondarwin replied:
Oops. Good spot. This was left over from some other ideas I was trying out.

gkalpak commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -33,141 +33,150 @@ export class DirectoryWalkerEntryPointFinder implements EntryPointFinder {
-        });
+      // This directory may contain entry-points of its own.
+      const childPaths = this.fs.readdir(absolutePath);
+      if (!isEntryPoint && childPaths.some(childPath => childPath.endsWith('.js'))) {
```
What if `childPath` is a directory (à la `zone.js`)? :grin:

petebacondarwin replied:
😭

petebacondarwin replied:
I added an additional check.

gkalpak commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -33,141 +33,150 @@ export class DirectoryWalkerEntryPointFinder implements EntryPointFinder {
+      const childPaths = this.fs.readdir(absolutePath);
+      if (!isEntryPoint && childPaths.some(childPath => childPath.endsWith('.js'))) {
+        // We do not consider non-entry-point directories that contain JS files as they are very
+        // unlikely to be containers for sub-entry-points.
```
Why is that?

petebacondarwin replied:
Because that is our experience - me and @alan-agius4 😸

gkalpak reviewed (changes requested):
A couple of minor comments/questions. Otherwise lgtm 👍

petebacondarwin commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -33,141 +33,144 @@ export class DirectoryWalkerEntryPointFinder implements EntryPointFinder {
+    const primaryEntryPoint =
+        getEntryPointInfo(this.fs, this.config, this.logger, sourceDirectory, sourceDirectory);
+
+    // If there is an entry-point but it is not a TypeScript one then exit
```
Strictly speaking it is INCOMPATIBLE which can mean either that the package.json is badly formed or that it is missing a valid `typings` property.

petebacondarwin replied:
I'll tweak the comment

petebacondarwin commented on packages/compiler-cli/ngcc/src/entry_point_finder/utils.ts:
```diff
@@ -78,3 +78,18 @@ function removeContainedPaths(value: AbsoluteFsPath, index: number, array: Absol
+ * @param log The function to call with the duration of the task
+ * @returns The result of calling `task`.
+ */
+export function trackDuration<T = void>(task: () => T, log: (duration: number) => void): T {
```
Sure

petebacondarwin committed:
fixup! refactor(ngcc): simplify `DirectoryWalkerEntryPointFinder`

petebacondarwin committed:
fixup! fix(ngcc): handle entry-points within container folders

petebacondarwin commented:
The perf tests are in (at least for my OS/X machine). I see no statistically significant difference between the `master` branch and this branch for running `ngcc`against the AIO `node_modules`.

**Time to walk directories**

run | master | PR
---- | -------- | ---
1    | 2.3 secs      | 1.0 secs
2    | 1.0 secs       | 0.9 secs
3    | 1.0 secs       | 1.3 secs

boskiv commented:
Can you tell me when it will be released?

petebacondarwin commented:
9.1.1 is due out on Wednesday
